### PR TITLE
Issue inline assembly stack warning if not balanced

### LINKED ISSUE #1173: Inline assembly: issue warning if stack height is different after the block
In most common cases the stack should be balanced after an inline assembly block:

```
assembly { 1 }
```

should raise a warning: `Inline assembly block leaves 1 item(s) on the stack`

### PR DESCRIPTION
**Pull request by axic**
Fixes #1173.

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -627,6 +628,22 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
 		}
 		return true;
 	});
+
+	int deposit = assembly.deposit();
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: I think this should be a feature of the inline assembly module itself. The place would be `void CodeTransform::operator()(assembly::Block const& _block)` - there is even a comment about stack height.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp]: Reworked. One downside I see is the same check is run 3 times now.

**Comment by axic** [comment]
@chriseth merge?

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -718,11 +718,10 @@ bool TypeChecker::visit(VariableDeclarationStatement const& _statement)
 		{
 			if (ref->dataStoredIn(DataLocation::Storage))
 			{
-				auto err = make_shared<Error>(Error::Type::Warning);
-				*err <<
-					errinfo_sourceLocation(varDecl.location()) <<
-					errinfo_comment("Uninitialized storage pointer. Did you mean '<type> memory " + varDecl.name() + "'?");
-				m_errors.push_back(err);
+				warning(
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:721]: This seems to be an unrelated change. Furthermore, why do you think a warning is better here?

File: libsolidity/codegen/ContractCompiler.cpp [hunk 3]
```diff
@@ -575,7 +575,7 @@ bool ContractCompiler::visit(InlineAssembly const& _inlineAssembly)
 			return true;
 		}
 	);
-	solAssert(errors.empty(), "Code generation for inline assembly with errors requested.");
+	solAssert(errors.empty() || Error::containsOnlyWarnings(errors), "Code generation for inline assembly with errors requested.");
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: The `errors.empty() ||` part is redundant.

File: libsolidity/inlineasm/AsmCodeGen.cpp [hunk 4]
```diff
@@ -20,14 +20,15 @@
  * Code-generating part of inline assembly.
  */
 
-#include <libsolidity/inlineasm/AsmCodeGen.h>
```
Comment (chriseth) [inline libsolidity/inlineasm/AsmCodeGen.cpp:23]: Putting a corresponding header first detects missing dependencies. I think we should follow this style here: http://stackoverflow.com/a/614322/5197297 , i.e.

own header first, then own project headers, general project headers, boost, std.

This order detects errors that occur because some missing dependencies of a header are included only because they are included by all users of that header.

File: libsolidity/inlineasm/AsmCodeGen.cpp [hunk 5]
```diff
@@ -214,9 +215,29 @@ class CodeTransform: public boost::static_visitor<>
 	{
 		size_t numVariables = m_state.variables.size();
 		std::for_each(_block.statements.begin(), _block.statements.end(), boost::apply_visitor(*this));
-		// pop variables
-		// we deliberately do not check stack height
+
 		m_state.assembly.setSourceLocation(_block.location);
+
+		// issue warnings for stack height discrepancies
+		int deposit = m_state.assembly.deposit();
```
Comment (chriseth) [inline libsolidity/inlineasm/AsmCodeGen.cpp]: I think you have to check that `m_state.assembly.deposit()` is not changed by line 216.

**Review by chriseth — CHANGES_REQUESTED** [review]
Unfortunately, I now remember why I did not check the stack height before:

Some constructs including jumps are only possible if this is not done. Basically, one could say that the stack offset counting is done incorrectly when a jump happens (this is not the case for jumpi).

So perhaps we should not issue the warning if the code generated by the block contains a jump opcode.

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp:721]: It is a warning already, just not using the internal helper.

`auto err = make_shared<Error>(Error::Type::Warning);`

File: libsolidity/inlineasm/AsmCodeGen.cpp (hunk 4 above)
Comment (axic) [reply libsolidity/inlineasm/AsmCodeGen.cpp:23]: It seemed from the CodingStandards.txt that the order was as this.  I've missed the `point b.`:

> b. The only exception to the above rule is the top of a .cpp file where its corresponding header should be located.

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:721]: Oh I'm sorry!

File: libsolidity/inlineasm/AsmCodeGen.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/inlineasm/AsmCodeGen.cpp:23]: Yes, I think we should change the coding standards. I'm not sure if the order has any impact on compilation performance.

File: libsolidity/inlineasm/AsmCodeGen.cpp (hunk 5 above)
Comment (axic) [reply libsolidity/inlineasm/AsmCodeGen.cpp]: You mean the part in `void operator()(assembly::Identifier const& _identifier)`?

File: libsolidity/codegen/ContractCompiler.cpp (hunk 3 above)
Comment (axic) [reply libsolidity/codegen/ContractCompiler.cpp]: Yes. The only part I disliked about `containsOnlyWarnings` that is shadows the fact if there are no warnings/errors at all. However that is not a problem here.

**Comment by axic** [comment]
> Some constructs including jumps are only possible if this is not done. 

Do you mean the warning produced would be invalid in that case? Because as above, warnings are not treated as errors anymore, so it should compile.

File: libsolidity/inlineasm/AsmCodeGen.cpp (hunk 5 above)
Comment (chriseth) [reply libsolidity/inlineasm/AsmCodeGen.cpp]: The block should not change the deposit. Usually, the deposit is already positive. I.e. store the deposit before the call to `for_all` and compare it here.

File: libsolidity/codegen/ContractCompiler.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp]: Then we should find a better name for that function `containsNoFatalErrors`? But yeah, different story.

**Comment by chriseth** [comment]
Ok, actually we need a way to modify the deposit from within inline assembly. The problem is that this value for deposit is used for the code that follows the inline assembly code. If it is wrong, then access to local variables is broken.

As a medium-term solution, let's always put this warning but perhaps add a comment that jumps may cause the value of deposit to be incorrect which also means that local variables access and other things will be broken in the code that follows the inline assembly.

File: libsolidity/inlineasm/AsmCodeGen.cpp (hunk 5 above)
Comment (axic) [reply libsolidity/inlineasm/AsmCodeGen.cpp]: Added.

**Review by chriseth — APPROVED** [review]
